File: apps/android/app/src/main/java/io/magicmobile/android/board/GameEmotes.kt
package io.magicmobile.android.board

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.compositionLocalOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.unit.dp
import io.magicmobile.android.game.GameSnapshot
import io.magicmobile.android.ui.GameAudio
import io.magicmobile.android.ui.GameSound
import io.magicmobile.android.ui.MagicPalette
import io.magicmobile.android.ui.SfDesign
import io.magicmobile.android.ui.SfImage
import io.magicmobile.android.ui.SfWeight
import io.magicmobile.android.ui.glow
import io.magicmobile.android.ui.rgb
import io.magicmobile.android.ui.sf
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.UUID
import kotlin.random.Random

/** Port of GameEmotes.swift. Quick chat: a fixed set of friendly lines, never free text. */
enum class GameEmote(val rawValue: String, val text: String, val symbol: String) {
    HELLO("hello", "Hello!", "hand.wave.fill"), WELL_PLAYED("wellPlayed", "Well played!", "hands.clap.fill"),
    THANKS("thanks", "Thanks!", "heart.fill"), OOPS("oops", "Oops!", "exclamationmark.bubble.fill"),
    WOW("wow", "Wow!", "sparkles"), GOOD_GAME("goodGame", "Good game!", "flag.checkered");

    /** What an AI opponent says back, when it answers at all. */
    val aiReply: GameEmote? get() = when (this) { HELLO -> HELLO; WELL_PLAYED -> THANKS; GOOD_GAME -> GOOD_GAME; WOW -> THANKS; else -> null }

    companion object { fun of(raw: String?): GameEmote? = entries.firstOrNull { it.rawValue == raw } }
}

/** Speech bubbles on the table, keyed by engine player ID, plus the way out to other players. */
class EmoteCenter(private val scope: CoroutineScope = MainScope()) {
    data class Bubble(val emote: GameEmote, val id: String = UUID.randomUUID().toString())

    val bubbles = mutableStateMapOf<String, Bubble>()
    var canSend by mutableStateOf(true); private set
    /** Online and Game Center-style games deliver your emote to the other players. */
    var send: ((GameEmote) -> Unit)? = null
    private val lastHeard = HashMap<String, Long>()

    fun show(emote: GameEmote, playerID: String) {
        val bubble = Bubble(emote)
        bubbles[playerID] = bubble
        GameAudio.play(GameSound.EMOTE)
        scope.launch { delay(2600); if (bubbles[playerID]?.id == bubble.id) bubbles.remove(playerID) }
    }

    /** You emote: show it, tell the other players, and let an AI opponent answer now and then. */
    fun say(emote: GameEmote, snapshot: GameSnapshot) {
        if (!canSend) return
        canSend = false
        scope.launch { delay(2000); canSend = true }
        show(emote, snapshot.viewerID)
        send?.invoke(emote)
        val bots = snapshot.players.filter { !snapshot.isViewer(it.playerId) && it.isHuman == false && !it.isOut }
        val reply = emote.aiReply
        val bot = bots.randomOrNull()
        if (reply != null && bot != null && Random.nextDouble() < 0.6) {
            scope.launch { delay((1100 + Random.nextDouble() * 900).toLong()); show(reply, bot.playerId) }
        }
    }

    /** Another player's emote, matched to their seat by the name shown at the table. */
    fun receive(emote: GameEmote, fromName: String, snapshot: GameSnapshot?) {
        snapshot ?: return
        val player = snapshot.players.firstOrNull { !snapshot.isViewer(it.playerId) && it.displayName == fromName } ?: return
        val now = System.currentTimeMillis()
        lastHeard[player.playerId]?.let { if (now - it < 1500) return }
        lastHeard[player.playerId] = now
        show(emote, player.playerId)
    }

    fun reset() { bubbles.clear(); lastHeard.clear() }
}

/** Set by on-device games (Swift `\.emoteCenter`). */
val LocalEmoteCenter = compositionLocalOf<EmoteCenter?> { null }

/** The bubble for one player, if they just emoted. */
@Composable
fun EmoteBubbleSlot(center: EmoteCenter, playerID: String, pointsUp: Boolean = false, modifier: Modifier = Modifier) {
    val bubble = center.bubbles[playerID]
    Box(modifier) {
        AnimatedVisibility(bubble != null,
            enter = if (BoardMotion.reduceMotion) fadeIn() else scaleIn(spring(0.62f, Spring.StiffnessMediumLow), 0.4f,
                TransformOrigin(0.5f, if (pointsUp) 0f else 1f)) + fadeIn(),
            exit = fadeOut()) {
            bubble?.let { EmoteBubble(it.emote, pointsUp) }
        }
    }
}

/** The latest opponent emote, named when it is not the opponent in focus. */
@Composable
fun OpponentEmoteSlot(center: EmoteCenter, snapshot: GameSnapshot, focusedID: String, modifier: Modifier = Modifier) {
    val opponents = snapshot.players.filter { !snapshot.isViewer(it.playerId) }
    val withBubbles = opponents.mapNotNull { player -> center.bubbles[player.playerId]?.let { player to it } }
    val latest = withBubbles.firstOrNull { it.first.playerId == focusedID } ?: withBubbles.firstOrNull()
    Box(modifier) {
        AnimatedVisibility(latest != null,
            enter = if (BoardMotion.reduceMotion) fadeIn() else scaleIn(spring(0.62f, Spring.StiffnessMediumLow), 0.4f, TransformOrigin(0.5f, 0f)) + fadeIn(),
            exit = fadeOut()) {
            latest?.let { (player, bubble) -> EmoteBubble(bubble.emote, name = if (player.playerId == focusedID) null else player.displayName) }
        }
    }
}

@Composable
fun EmoteBubble(emote: GameEmote, pointsUp: Boolean = false, name: String? = null) {
    val shape = RoundedCornerShape(14.dp)
    Row(Modifier.glow(Color.Black.copy(alpha = 0.45f), 8.dp, 14.dp)
        .background(Brush.verticalGradient(listOf(rgb(1.0, 0.98, 0.92), rgb(0.93, 0.87, 0.74))), shape)
        .border(1.5.dp, MagicPalette.antiqueGold, shape).padding(horizontal = 12.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
        SfImage(emote.symbol, MagicPalette.oxblood, 15.dp)
        if (name != null) Text("$name:", color = MagicPalette.oxblood, style = sf(13f, SfWeight.heavy, SfDesign.ROUNDED), softWrap = false)
        Text(emote.text, color = rgb(0.16, 0.11, 0.08), style = sf(15f, SfWeight.heavy, SfDesign.ROUNDED), softWrap = false)
    }
}

/** Emote choices shown when you tap your own life orb. */
@Composable
fun EmotePicker(center: EmoteCenter, snapshot: GameSnapshot, done: () -> Unit) {
    Column(Modifier.width(300.dp).background(MagicPalette.iron.copy(alpha = 0.97f)).padding(14.dp), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Text("QUICK CHAT", color = MagicPalette.antiqueGold, style = sf(11f, SfWeight.black, tracking = 1.6f))
        for (row in GameEmote.entries.chunked(2)) {
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                for (emote in row) {
                    CompactActionButton(isPrimary = false, enabled = center.canSend, modifier = Modifier.weight(1f),
                        onClick = { center.say(emote, snapshot); done() }) {
                        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.Center, verticalAlignment = Alignment.CenterVertically) {
                            SfImage(emote.symbol, Color.White, 15.dp)
                            Text(" ${emote.text}", style = sf(14f, SfWeight.heavy, SfDesign.ROUNDED), color = Color.White)
                        }
                    }
                }
            }
        }
    }
}

File: apps/ios/MagicMobile/GameEmotes.swift
import SwiftUI

/// Quick chat: a fixed set of friendly lines, never free text.
enum GameEmote: String, CaseIterable, Identifiable {
    case hello, wellPlayed, thanks, oops, wow, goodGame

    var id: String { rawValue }

    var text: String {
        switch self {
        case .hello: return String(localized: "Hello!")
        case .wellPlayed: return String(localized: "Well played!")
        case .thanks: return String(localized: "Thanks!")
        case .oops: return String(localized: "Oops!")
        case .wow: return String(localized: "Wow!")
        case .goodGame: return String(localized: "Good game!")
        }
    }

    var symbol: String {
        switch self {
        case .hello: return "hand.wave.fill"
        case .wellPlayed: return "hands.clap.fill"
        case .thanks: return "heart.fill"
        case .oops: return "exclamationmark.bubble.fill"
        case .wow: return "sparkles"
        case .goodGame: return "flag.checkered"
        }
    }

    /// What an AI opponent says back, when it answers at all.
    var aiReply: GameEmote? {
        switch self {
        case .hello: return .hello
        case .wellPlayed: return .thanks
        case .goodGame: return .goodGame
        case .wow: return .thanks
        case .oops, .thanks: return nil
        }
    }
}

/// Speech bubbles on the table, keyed by engine player ID, plus the way out to other players.
@MainActor
final class EmoteCenter: ObservableObject {
    struct Bubble: Equatable {
        let id = UUID()
        let emote: GameEmote
    }

    @Published private(set) var bubbles: [String: Bubble] = [:]
    @Published private(set) var canSend = true
    /// Game Center: delivers your emote to the other players.
    var send: ((GameEmote) -> Void)?
    private var lastHeard: [String: Date] = [:]

    func show(_ emote: GameEmote, from playerID: String) {
        let bubble = Bubble(emote: emote)
        bubbles[playerID] = bubble
        GameAudio.shared.play(.emote)
        Task { @MainActor [weak self] in
            try? await Task.sleep(for: .seconds(2.6))
            if self?.bubbles[playerID]?.id == bubble.id { self?.bubbles[playerID] = nil }
        }
    }

    /// You emote: show it, tell the other players, and let an AI opponent answer now and then.
    func say(_ emote: GameEmote, in snapshot: GameSnapshot) {
        guard canSend else { return }
        canSend = false
        Task { @MainActor [weak self] in
            try? await Task.sleep(for: .seconds(2))
            self?.canSend = true
        }
        show(emote, from: snapshot.viewerID)
        send?(emote)
        let bots = snapshot.players.filter { !snapshot.isViewer($0.playerId) && $0.isHuman == false && !$0.isOut }
        if let reply = emote.aiReply, let bot = bots.randomElement(), Double.random(in: 0..<1) < 0.6 {
            Task { @MainActor [weak self] in
                try? await Task.sleep(for: .seconds(Double.random(in: 1.1...2.0)))
                self?.show(reply, from: bot.playerId)
            }
        }
    }

    /// A Game Center player's emote, matched to their seat by the name shown at the table.
    func receive(_ emote: GameEmote, fromName name: String, in snapshot: GameSnapshot?) {
        guard let snapshot, let player = snapshot.players.first(where: {
            !snapshot.isViewer($0.playerId) && $0.displayName == name
        }) else { return }
        if let last = lastHeard[player.playerId], Date().timeIntervalSince(last) < 1.5 { return }
        lastHeard[player.playerId] = Date()
        show(emote, from: player.playerId)
    }

    func reset() { bubbles = [:]; lastHeard = [:] }
}

private struct EmoteCenterKey: EnvironmentKey { static let defaultValue: EmoteCenter? = nil }

extension EnvironmentValues {
    /// Set by on-device games; the hosted board has no quick chat.
    var emoteCenter: EmoteCenter? {
        get { self[EmoteCenterKey.self] }
        set { self[EmoteCenterKey.self] = newValue }
    }
}

/// The bubble for one player, if they just emoted.
struct EmoteBubbleSlot: View {
    @ObservedObject var center: EmoteCenter
    let playerID: String
    var pointsUp = false
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        ZStack {
            if let bubble = center.bubbles[playerID] {
                EmoteBubble(emote: bubble.emote, pointsUp: pointsUp)
                    .id(bubble.id)
                    .transition(reduceMotion ? .opacity : .scale(scale: 0.4, anchor: pointsUp ? .top : .bottom).combined(with: .opacity))
            }
        }
        .animation(reduceMotion ? nil : .spring(response: 0.32, dampingFraction: 0.62), value: center.bubbles[playerID])
        .allowsHitTesting(false)
    }
}

/// The latest opponent emote, named when it is not the opponent in focus.
struct OpponentEmoteSlot: View {
    @ObservedObject var center: EmoteCenter
    let snapshot: GameSnapshot
    let focusedID: String
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    private var latest: (player: PlayerGameState, bubble: EmoteCenter.Bubble)? {
        snapshot.players.filter { !snapshot.isViewer($0.playerId) }
            .compactMap { player in center.bubbles[player.playerId].map { (player, $0) } }
            .first { $0.player.playerId == focusedID } ?? snapshot.players.filter { !snapshot.isViewer($0.playerId) }
            .compactMap { player in center.bubbles[player.playerId].map { (player, $0) } }.first
    }

    var body: some View {
        ZStack {
            if let latest {
                EmoteBubble(emote: latest.bubble.emote,
                            name: latest.player.playerId == focusedID ? nil : latest.player.displayName)
                    .id(latest.bubble.id)
                    .transition(reduceMotion ? .opacity : .scale(scale: 0.4, anchor: .top).combined(with: .opacity))
            }
        }
        .animation(reduceMotion ? nil : .spring(response: 0.32, dampingFraction: 0.62), value: latest?.bubble)
        .allowsHitTesting(false)
    }
}

struct EmoteBubble: View {
    let emote: GameEmote
    var pointsUp = false
    var name: String? = nil

    var body: some View {
        HStack(spacing: 6) {
            Image(systemName: emote.symbol).font(.system(size: 13, weight: .bold))
                .foregroundStyle(MagicPalette.oxblood)
            if let name {
                Text("\(name):")
                    .font(.system(size: 13, weight: .heavy, design: .rounded))
                    .foregroundStyle(MagicPalette.oxblood)
                    .fixedSize()
            }
            Text(emote.text)
                .font(.system(size: 15, weight: .heavy, design: .rounded))
                .foregroundStyle(Color(red: 0.16, green: 0.11, blue: 0.08))
                .fixedSize()
        }
        .padding(.horizontal, 12).padding(.vertical, 8)
        .background(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .fill(LinearGradient(colors: [Color(red: 1, green: 0.98, blue: 0.92), Color(red: 0.93, green: 0.87, blue: 0.74)],
                                     startPoint: .top, endPoint: .bottom))
        )
        .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous).stroke(MagicPalette.antiqueGold, lineWidth: 1.5))
        .shadow(color: .black.opacity(0.45), radius: 8, y: 4)
        .accessibilityElement(children: .combine)
        .accessibilityAddTraits(.updatesFrequently)
    }
}

/// Emote choices shown when you tap your own life orb.
struct EmotePicker: View {
    @ObservedObject var center: EmoteCenter
    let snapshot: GameSnapshot
    let done: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("QUICK CHAT")
                .font(.system(size: 11, weight: .black)).tracking(1.6)
                .foregroundStyle(MagicPalette.antiqueGold)
            LazyVGrid(columns: [GridItem(.flexible(), spacing: 8), GridItem(.flexible(), spacing: 8)], spacing: 8) {
                ForEach(GameEmote.allCases) { emote in
                    Button {
                        center.say(emote, in: snapshot)
                        done()
                    } label: {
                        Label(emote.text, systemImage: emote.symbol)
                            .font(.system(size: 14, weight: .heavy, design: .rounded))
                            .frame(maxWidth: .infinity, minHeight: 44)
                    }
                    .buttonStyle(CompactActionButtonStyle(isPrimary: false))
                    .disabled(!center.canSend)
                    .accessibilityIdentifier("board.emote.\(emote.rawValue)")
                }
            }
        }
        .padding(14)
        .frame(width: 300)
        .background(MagicPalette.iron.opacity(0.97))
    }
}
